[BUG] Java process does not get killed when data prepper itself terminates

**Describe the bug**
The running java process does not get killed when data prepper itself terminates. This is problematic when running data prepper inside of a docker container using kubernetes HorizontalPodAutoscaler to control data prepper. Right now data prepper will terminate for some reason however the container/pod does not terminate and will stick around as a zombie pod- holding onto resources but unable to function. I believe this is because PID 1 (the java cmd running data prepper) does not get terminated when the data-prepper process stops. 

**To Reproduce**
Really you just need data prepper to try to use more resources than is available in a container/pod. Here is one way
1. Run data prepper in kubernetes
1. Set the pod resource limit to something low like [cpu 100m](https://kubernetes.io/docs/concepts/configuration/manage-resources-containers/#example-1) to force it to run into resource issues
1. Configure data prepper to accept data
1. Fire a lot of data at data prepper really fast (One way is to build up a buffer in your source and then try to send it to data prepper all at once) What I did was force restart data prepper a few times while simultaneously generated a bunch of source data trying to send to data prepper. Then data prepper gets overwhelmed, tries to get more resources, throws an error and terminates

Sample error
```
2023-04-03T17:18:30,799 [pool-7-thread-83] ERROR com.amazon.dataprepper.plugins.source.otellogs.OTelLogsGrpcService - Failed to write the request of size 2380052 due to:
java.util.concurrent.TimeoutException: Pipeline [log-pipeline] - Buffer does not have enough capacity left for the size of records: 2611, timed out waiting for slots.
	at org.opensearch.dataprepper.plugins.buffer.blockingbuffer.BlockingBuffer.doWriteAll(BlockingBuffer.java:123) ~[blocking-buffer-2.1.0.jar:?]
	at org.opensearch.dataprepper.model.buffer.AbstractBuffer.writeAll(AbstractBuffer.java:100) ~[data-prepper-api-2.1.0.jar:?]
	at org.opensearch.dataprepper.plugins.MultiBufferDecorator.writeAll(MultiBufferDecorator.java:39) ~[data-prepper-core-2.1.0.jar:?]
	at com.amazon.dataprepper.plugins.source.otellogs.OTelLogsGrpcService.processRequest(OTelLogsGrpcService.java:106) ~[otel-logs-source-2.1.0.jar:?]
	at com.amazon.dataprepper.plugins.source.otellogs.OTelLogsGrpcService.lambda$export$0(OTelLogsGrpcService.java:80) ~[otel-logs-source-2.1.0.jar:?]
	at io.micrometer.core.instrument.composite.CompositeTimer.record(CompositeTimer.java:141) ~[micrometer-core-1.10.3.jar:1.10.3]
	at com.amazon.dataprepper.plugins.source.otellogs.OTelLogsGrpcService.export(OTelLogsGrpcService.java:80) ~[otel-logs-source-2.1.0.jar:?]
	at io.opentelemetry.proto.collector.logs.v1.LogsServiceGrpc$MethodHandlers.invoke(LogsServiceGrpc.java:246) ~[opentelemetry-proto-0.16.0-alpha.jar:0.16.0]
	at io.grpc.stub.ServerCalls$UnaryServerCallHandler$UnaryServerCallListener.onHalfClose(ServerCalls.java:182) ~[grpc-stub-1.49.1.jar:1.49.1]
	at com.linecorp.armeria.server.grpc.AbstractServerCall.invokeOnMessage(AbstractServerCall.java:384) ~[armeria-grpc-1.22.1.jar:?]
	at com.linecorp.armeria.server.grpc.AbstractServerCall.lambda$onRequestMessage$2(AbstractServerCall.java:348) ~[armeria-grpc-1.22.1.jar:?]
	at com.linecorp.armeria.internal.shaded.guava.util.concurrent.SequentialExecutor$1.run(SequentialExecutor.java:123) ~[armeria-1.22.1.jar:?]
	at com.linecorp.armeria.internal.shaded.guava.util.concurrent.SequentialExecutor$QueueWorker.workOnQueue(SequentialExecutor.java:235) ~[armeria-1.22.1.jar:?]
	at com.linecorp.armeria.internal.shaded.guava.util.concurrent.SequentialExecutor$QueueWorker.run(SequentialExecutor.java:180) ~[armeria-1.22.1.jar:?]
	at com.linecorp.armeria.common.RequestContext.lambda$makeContextAware$3(RequestContext.java:566) ~[armeria-1.22.1.jar:?]
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:539) ~[?:?]
	at java.util.concurrent.FutureTask.run(FutureTask.java:264) ~[?:?]
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:304) ~[?:?]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136) ~[?:?]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635) ~[?:?]
	at java.lang.Thread.run(Thread.java:833) ~[?:?]
2023-04-03T17:18:30,800 [log-pipeline-sink-worker-4-thread-1] INFO  org.opensearch.dataprepper.pipeline.Pipeline - Pipeline [log-pipeline] - Shutting down process workers
2023-04-03T17:19:01,806 [log-pipeline-sink-worker-4-thread-1] WARN  org.opensearch.dataprepper.pipeline.Pipeline - Pipeline [log-pipeline] - Workers did not terminate in time, forcing termination
2023-04-03T17:19:05,278 [log-pipeline-processor-worker-3-thread-1] ERROR org.opensearch.dataprepper.pipeline.common.FutureHelper - FutureTask is interrupted or timed out
2023-04-03T17:19:06,287 [log-pipeline-processor-worker-3-thread-1] INFO  org.opensearch.dataprepper.pipeline.ProcessWorker - Processor shutdown phase 1 complete.
2023-04-03T17:19:07,300 [log-pipeline-processor-worker-3-thread-1] INFO  org.opensearch.dataprepper.pipeline.ProcessWorker - Beginning processor shutdown phase 2, iterating until buffers empty.
2023-04-03T17:19:07,300 [log-pipeline-sink-worker-4-thread-1] INFO  org.opensearch.dataprepper.pipeline.Pipeline - Pipeline [log-pipeline] - Shutting down process workers
Exception in thread "HTTP-Dispatcher" Exception in thread "I/O dispatcher 1" java.lang.OutOfMemoryError: Java heap space
java.lang.OutOfMemoryError: Java heap space
2023-04-03T17:20:00,693 [log-pipeline-sink-worker-4-thread-1] WARN  org.opensearch.dataprepper.pipeline.Pipeline - Pipeline [log-pipeline] - Workers did not terminate in time, forcing termination
2023-04-03T17:20:00,693 [metrics-pipeline-processor-worker-5-thread-1] ERROR org.opensearch.dataprepper.pipeline.common.PipelineThreadPoolExecutor - Pipeline [metrics-pipeline] process worker encountered a fatal exception, cannot proceed further
java.util.concurrent.ExecutionException: java.lang.OutOfMemoryError: Java heap space: failed reallocation of scalar replaced objects
	at java.util.concurrent.FutureTask.report(FutureTask.java:122) ~[?:?]
	at java.util.concurrent.FutureTask.get(FutureTask.java:191) ~[?:?]
	at org.opensearch.dataprepper.pipeline.common.PipelineThreadPoolExecutor.afterExecute(PipelineThreadPoolExecutor.java:70) ~[data-prepper-core-2.1.0.jar:?]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1137) ~[?:?]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635) ~[?:?]
	at java.lang.Thread.run(Thread.java:833) ~[?:?]
Caused by: java.lang.OutOfMemoryError: Java heap space: failed reallocation of scalar replaced objects
2023-04-03T17:20:00,693 [metrics-pipeline-processor-worker-5-thread-1] INFO  org.opensearch.dataprepper.pipeline.Pipeline - Pipeline [metrics-pipeline] - Received shutdown signal with processor shutdown timeout PT30S and sink shutdown timeout PT30S. Initiating the shutdown process
2023-04-03T17:20:00,693 [log-pipeline-processor-worker-3-thread-1] ERROR org.opensearch.dataprepper.pipeline.ProcessWorker - Encountered exception during pipeline log-pipeline processing
java.util.concurrent.RejectedExecutionException: Task java.util.concurrent.FutureTask@416a9cb0[Not completed, task = java.util.concurrent.Executors$RunnableAdapter@706b5ab7[Wrapped task = org.opensearch.dataprepper.pipeline.Pipeline$$Lambda$1099/0x00000008014068d8@6dccc3d5]] rejected from org.opensearch.dataprepper.pipeline.common.PipelineThreadPoolExecutor@6f8299f9[Shutting down, pool size = 1, active threads = 1, queued tasks = 1, completed tasks = 109]
	at java.util.concurrent.ThreadPoolExecutor$AbortPolicy.rejectedExecution(ThreadPoolExecutor.java:2065) ~[?:?]
	at java.util.concurrent.ThreadPoolExecutor.reject(ThreadPoolExecutor.java:833) ~[?:?]
	at java.util.concurrent.ThreadPoolExecutor.execute(ThreadPoolExecutor.java:1365) ~[?:?]
	at java.util.concurrent.AbstractExecutorService.submit(AbstractExecutorService.java:134) ~[?:?]
	at org.opensearch.dataprepper.pipeline.Pipeline.lambda$publishToSinks$4(Pipeline.java:262) ~[data-prepper-core-2.1.0.jar:?]
	at org.opensearch.dataprepper.pipeline.router.DataFlowComponentRouter.route(DataFlowComponentRouter.java:45) ~[data-prepper-core-2.1.0.jar:?]
	at org.opensearch.dataprepper.pipeline.router.Router.route(Router.java:42) ~[data-prepper-core-2.1.0.jar:?]
	at org.opensearch.dataprepper.pipeline.Pipeline.publishToSinks(Pipeline.java:261) ~[data-prepper-core-2.1.0.jar:?]
	at org.opensearch.dataprepper.pipeline.ProcessWorker.postToSink(ProcessWorker.java:117) ~[data-prepper-core-2.1.0.jar:?]
	at org.opensearch.dataprepper.pipeline.ProcessWorker.doRun(ProcessWorker.java:98) ~[data-prepper-core-2.1.0.jar:?]
	at org.opensearch.dataprepper.pipeline.ProcessWorker.run(ProcessWorker.java:52) ~[data-prepper-core-2.1.0.jar:?]
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:539) ~[?:?]
	at java.util.concurrent.FutureTask.run(FutureTask.java:264) ~[?:?]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136) ~[?:?]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635) ~[?:?]
	at java.lang.Thread.run(Thread.java:833) ~[?:?]
2023-04-03T17:20:00,694 [metrics-pipeline-processor-worker-5-thread-1] INFO  org.opensearch.dataprepper.plugins.source.otelmetrics.OTelMetricsSource - Stopped otel_metrics_source.
2023-04-03T17:20:00,694 [metrics-pipeline-processor-worker-5-thread-1] INFO  org.opensearch.dataprepper.pipeline.Pipeline - Pipeline [metrics-pipeline] - Shutting down process workers
2023-04-03T17:20:00,694 [metrics-pipeline-processor-worker-5-thread-1] INFO  org.opensearch.dataprepper.pipeline.Pipeline - Pipeline [metrics-pipeline] - Encountered interruption terminating the pipeline execution, Attempting to force the termination
2023-04-03T17:20:00,694 [metrics-pipeline-processor-worker-5-thread-1] INFO  org.opensearch.dataprepper.pipeline.Pipeline - Pipeline [metrics-pipeline] - Shutting down process workers
2023-04-03T17:20:00,694 [pool-4-thread-1] ERROR org.apache.http.impl.nio.client.CloseableHttpAsyncClientBase$1 - I/O reactor terminated abnormally
org.apache.http.nio.reactor.IOReactorException: I/O dispatch worker terminated abnormally
	at org.apache.http.impl.nio.reactor.AbstractMultiworkerIOReactor.execute(AbstractMultiworkerIOReactor.java:359) ~[httpcore-nio-4.4.15.jar:4.4.15]
	at org.apache.http.impl.nio.conn.PoolingNHttpClientConnectionManager.execute(PoolingNHttpClientConnectionManager.java:221) ~[httpasyncclient-4.1.5.jar:4.1.5]
	at org.apache.http.impl.nio.client.CloseableHttpAsyncClientBase$1.run(CloseableHttpAsyncClientBase.java:64) ~[httpasyncclient-4.1.5.jar:4.1.5]
	at java.lang.Thread.run(Thread.java:833) ~[?:?]
Caused by: java.lang.OutOfMemoryError: Java heap space
```

**Expected behavior**
PID 1 to be terminated, and then the container/pod to terminate because PID 1 (the init cmd for the container) was terminated. 

I ran this after it crashed & you can see PID 1 was not terminated
```
data-prepper-775db9b7b6-msg2v:/usr/share/data-prepper# ps -A
PID   USER     TIME  COMMAND
    1 root      5:23 java -Ddata-prepper.dir=/usr/share/data-prepper -Dlog4j.configurationFile=/usr/share/data-prepper/config/log4j2-rolling.properties -cp /u
  482 root      0:00 sh -c clear; (bash || ash || sh)
  489 root      0:00 bash
  490 root      0:00 ps -A
```

**Environment (please complete the following information):**
- Docker
- Kubernetes
- Data prepper v2.1.0

**Additional context**
Note: data prepper does not crash when the resource cpu/mem limit is gone and I can view the resources used spike way over my preset limit if I run the same test.

It appears that we have seen a similar issue when exceptions are thrown in the OpenSearch sink.


```
2023-04-05T14:29:51.247 [log-pipeline-sink-worker-2-thread-2] ERROR org.opensearch.dataprepper.pipeline.common.PipelineThreadPoolExecutor - Pipeline [log-pipeline] process worker encountered a fatal exception, cannot proceed further
java.util.concurrent.ExecutionException: java.lang.NullPointerException
at java.util.concurrent.FutureTask.report(FutureTask.java:122) ~[?:?]
at java.util.concurrent.FutureTask.get(FutureTask.java:191) ~[?:?]
at org.opensearch.dataprepper.pipeline.common.PipelineThreadPoolExecutor.afterExecute(PipelineThreadPoolExecutor.java:70) ~[data-prepper-core-2.2.0-SNAPSHOT.jar:?]
at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1129) ~[?:?]
at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628) ~[?:?]
at java.lang.Thread.run(Thread.java:829) [?:?]
Caused by: java.lang.NullPointerException
at org.opensearch.dataprepper.plugins.sink.opensearch.BulkRetryStrategy.handleRetry(BulkRetryStrategy.java:195) ~[opensearch-2.2.0-SNAPSHOT.jar:?]
at org.opensearch.dataprepper.plugins.sink.opensearch.BulkRetryStrategy.execute(BulkRetryStrategy.java:145) ~[opensearch-2.2.0-SNAPSHOT.jar:?]
at org.opensearch.dataprepper.plugins.sink.opensearch.OpenSearchSink.lambda$flushBatch$6(OpenSearchSink.java:243) ~[opensearch-2.2.0-SNAPSHOT.jar:?]
at io.micrometer.core.instrument.composite.CompositeTimer.record(CompositeTimer.java:141) ~[micrometer-core-1.10.3.jar:1.10.3]
at org.opensearch.dataprepper.plugins.sink.opensearch.OpenSearchSink.flushBatch(OpenSearchSink.java:240) ~[opensearch-2.2.0-SNAPSHOT.jar:?]
at org.opensearch.dataprepper.plugins.sink.opensearch.OpenSearchSink.doOutput(OpenSearchSink.java:228) ~[opensearch-2.2.0-SNAPSHOT.jar:?]
at org.opensearch.dataprepper.model.sink.AbstractSink.lambda$output$0(AbstractSink.java:54) ~[data-prepper-api-2.2.0-SNAPSHOT.jar:?]
at io.micrometer.core.instrument.composite.CompositeTimer.record(CompositeTimer.java:141) ~[micrometer-core-1.10.3.jar:1.10.3]
at org.opensearch.dataprepper.model.sink.AbstractSink.output(AbstractSink.java:54) ~[data-prepper-api-2.2.0-SNAPSHOT.jar:?]
at org.opensearch.dataprepper.pipeline.Pipeline.lambda$publishToSinks$3(Pipeline.java:286) ~[data-prepper-core-2.2.0-SNAPSHOT.jar:?]
at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515) ~[?:?]
at java.util.concurrent.FutureTask.run(FutureTask.java:264) ~[?:?]
at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128) ~[?:?]
... 2 more
2023-04-05T14:29:51.253 [log-pipeline-processor-worker-1-thread-1] ERROR org.opensearch.dataprepper.pipeline.common.FutureHelper - FutureTask failed due to:
java.util.concurrent.ExecutionException: java.lang.NullPointerException
at java.util.concurrent.FutureTask.report(FutureTask.java:122) ~[?:?]
at java.util.concurrent.FutureTask.get(FutureTask.java:191) ~[?:?]
at org.opensearch.dataprepper.pipeline.common.FutureHelper.awaitFuturesIndefinitely(FutureHelper.java:29) ~[data-prepper-core-2.2.0-SNAPSHOT.jar:?]
at org.opensearch.dataprepper.pipeline.ProcessWorker.postToSink(ProcessWorker.java:118) ~[data-prepper-core-2.2.0-SNAPSHOT.jar:?]
at org.opensearch.dataprepper.pipeline.ProcessWorker.doRun(ProcessWorker.java:98) ~[data-prepper-core-2.2.0-SNAPSHOT.jar:?]
at org.opensearch.dataprepper.pipeline.ProcessWorker.run(ProcessWorker.java:45) ~[data-prepper-core-2.2.0-SNAPSHOT.jar:?]
at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515) ~[?:?]
at java.util.concurrent.FutureTask.run(FutureTask.java:264) ~[?:?]
at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128) ~[?:?]
at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628) ~[?:?]
at java.lang.Thread.run(Thread.java:829) [?:?]
Caused by: java.lang.NullPointerException
at org.opensearch.dataprepper.plugins.sink.opensearch.BulkRetryStrategy.handleRetry(BulkRetryStrategy.java:195) ~[opensearch-2.2.0-SNAPSHOT.jar:?]
at org.opensearch.dataprepper.plugins.sink.opensearch.BulkRetryStrategy.execute(BulkRetryStrategy.java:145) ~[opensearch-2.2.0-SNAPSHOT.jar:?]
at org.opensearch.dataprepper.plugins.sink.opensearch.OpenSearchSink.lambda$flushBatch$6(OpenSearchSink.java:243) ~[opensearch-2.2.0-SNAPSHOT.jar:?]
at io.micrometer.core.instrument.composite.CompositeTimer.record(CompositeTimer.java:141) ~[micrometer-core-1.10.3.jar:1.10.3]
at org.opensearch.dataprepper.plugins.sink.opensearch.OpenSearchSink.flushBatch(OpenSearchSink.java:240) ~[opensearch-2.2.0-SNAPSHOT.jar:?]
at org.opensearch.dataprepper.plugins.sink.opensearch.OpenSearchSink.doOutput(OpenSearchSink.java:228) ~[opensearch-2.2.0-SNAPSHOT.jar:?]
at org.opensearch.dataprepper.model.sink.AbstractSink.lambda$output$0(AbstractSink.java:54) ~[data-prepper-api-2.2.0-SNAPSHOT.jar:?]
at io.micrometer.core.instrument.composite.CompositeTimer.record(CompositeTimer.java:141) ~[micrometer-core-1.10.3.jar:1.10.3]
at org.opensearch.dataprepper.model.sink.AbstractSink.output(AbstractSink.java:54) ~[data-prepper-api-2.2.0-SNAPSHOT.jar:?]
at org.opensearch.dataprepper.pipeline.Pipeline.lambda$publishToSinks$3(Pipeline.java:286) ~[data-prepper-core-2.2.0-SNAPSHOT.jar:?]
... 5 more
```

--- a/data-prepper-core/src/integrationTest/java/org/opensearch/dataprepper/test/framework/DataPrepperTestRunner.java
+++ b/data-prepper-core/src/integrationTest/java/org/opensearch/dataprepper/test/framework/DataPrepperTestRunner.java
@@ -5,6 +5,7 @@
 
 package org.opensearch.dataprepper.test.framework;
 
+import org.opensearch.dataprepper.AbstractContextManager;
 import org.opensearch.dataprepper.DataPrepper;
 import org.opensearch.dataprepper.parser.config.FileStructurePathProvider;
 import org.opensearch.dataprepper.plugins.InMemorySinkAccessor;
@@ -25,13 +26,11 @@ import javax.annotation.Nullable;
 public class DataPrepperTestRunner {
     private static final Logger LOG = LoggerFactory.getLogger(DataPrepperTestRunner.class);
     private static final String BASE_PATH = "src/integrationTest/resources/org/opensearch/dataprepper";
-    private static final String BASE_DATA_PREPPER_PACKAGE = "org.opensearch.dataprepper";
-    private static final String EXPRESSION_PACKAGE = BASE_DATA_PREPPER_PACKAGE + ".expression";
-    private final AnnotationConfigApplicationContext coreApplicationContext;
     private final String dataPrepperConfigFile;
     private final String pipelinesDirectoryOrFile;
     private final InMemorySourceAccessor inMemorySourceAccessor;
     private final InMemorySinkAccessor inMemorySinkAccessor;
+    private final TestContextManager contextManager;
 
     private DataPrepperTestRunner(final Builder builder) {
         dataPrepperConfigFile = builder.dataPrepperConfigFile;
@@ -41,18 +40,7 @@ public class DataPrepperTestRunner {
         inMemorySinkAccessor = new InMemorySinkAccessor();
         LOG.info("created in memory source accessor {}", inMemorySourceAccessor);
 
-        final AnnotationConfigApplicationContext publicApplicationContext = new AnnotationConfigApplicationContext();
-        publicApplicationContext.scan(EXPRESSION_PACKAGE);
-        publicApplicationContext.registerBean(InMemorySourceAccessor.class, () -> inMemorySourceAccessor);
-        publicApplicationContext.registerBean(InMemorySinkAccessor.class, () -> inMemorySinkAccessor);
-        publicApplicationContext.refresh();
-
-        coreApplicationContext = new AnnotationConfigApplicationContext();
-        coreApplicationContext.setParent(publicApplicationContext);
-        coreApplicationContext.registerBean(FileStructurePathProvider.class, TestFileStructurePathProvider::new);
-        coreApplicationContext.scan(BASE_DATA_PREPPER_PACKAGE);
-
-        coreApplicationContext.refresh();
+        contextManager = new TestContextManager();
         LOG.info("Started Data Prepper Application context for testing.");
     }
 
@@ -68,7 +56,7 @@ public class DataPrepperTestRunner {
      * Starts the Data Prepper test instance.
      */
     public void start() {
-        coreApplicationContext.getBean(DataPrepper.class)
+        contextManager.getDataPrepperBean()
                 .execute();
     }
 
@@ -76,9 +64,8 @@ public class DataPrepperTestRunner {
      * Stops the Data Prepper test instance.
      */
     public void stop() {
-        final DataPrepper dataPrepper = coreApplicationContext.getBean(DataPrepper.class);
+        final DataPrepper dataPrepper = contextManager.getDataPrepperBean();
         dataPrepper.shutdown();
-        dataPrepper.shutdownServers();
     }
 
     /**
@@ -134,6 +121,19 @@ public class DataPrepperTestRunner {
 
         public DataPrepperTestRunner build() {
             return new DataPrepperTestRunner(this);
+        }
+    }
+
+    private class TestContextManager extends AbstractContextManager {
+
+        protected void preRefreshPublicApplicationContext(final AnnotationConfigApplicationContext publicApplicationContext) {
+            publicApplicationContext.registerBean(InMemorySourceAccessor.class, () -> inMemorySourceAccessor);
+            publicApplicationContext.registerBean(InMemorySinkAccessor.class, () -> inMemorySinkAccessor);
+
+        }
+
+        protected void preRefreshCoreApplicationContext(final AnnotationConfigApplicationContext coreApplicationContext) {
+            coreApplicationContext.registerBean(FileStructurePathProvider.class, TestFileStructurePathProvider::new);
         }
     }
 }

--- a/data-prepper-core/src/main/java/org/opensearch/dataprepper/AbstractContextManager.java
+++ b/data-prepper-core/src/main/java/org/opensearch/dataprepper/AbstractContextManager.java
@@ -1,0 +1,98 @@
+/*
+ * Copyright OpenSearch Contributors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package org.opensearch.dataprepper;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.context.annotation.AnnotationConfigApplicationContext;
+
+/**
+ * Base class for managing context. It provides hooks for inheritors to control the context to some degree.
+ * <p>Creates Spring {@link org.springframework.context.ApplicationContext} hierarchy for Dependency Injection with limited visibility.</p>
+ * <p>
+ *     Application Context Hierarchy
+ *         Public Application Context<br>
+ *         ├─ Core Application Context<br>
+ *         ├─ Shared Plugin Application Context<br>
+ *             ├─ Plugin Isolated Application Context<br>
+ * </p>
+ * @since 2.2
+ */
+public abstract class AbstractContextManager {
+    private static final Logger LOG = LoggerFactory.getLogger(AbstractContextManager.class);
+    private static final String BASE_DATA_PREPPER_PACKAGE = "org.opensearch.dataprepper";
+    private static final String EXPRESSION_PACKAGE = BASE_DATA_PREPPER_PACKAGE + ".expression";
+
+    private final AnnotationConfigApplicationContext publicApplicationContext;
+    private final AnnotationConfigApplicationContext coreApplicationContext;
+    private DataPrepper dataPrepper;
+
+    protected AbstractContextManager() {
+        publicApplicationContext = new AnnotationConfigApplicationContext();
+        coreApplicationContext = new AnnotationConfigApplicationContext();
+    }
+
+    /**
+     * @since 1.3
+     * Retrieves the instance of {@link DataPrepper} singleton
+     * @return {@link DataPrepper} instance
+     */
+    public DataPrepper getDataPrepperBean() {
+        if(dataPrepper == null) {
+            start();
+        }
+        return dataPrepper;
+    }
+
+    private void start() {
+        publicApplicationContext.scan(EXPRESSION_PACKAGE);
+        preRefreshPublicApplicationContext(publicApplicationContext);
+
+        publicApplicationContext.refresh();
+        coreApplicationContext.setParent(publicApplicationContext);
+        coreApplicationContext.scan(BASE_DATA_PREPPER_PACKAGE);
+        preRefreshCoreApplicationContext(coreApplicationContext);
+
+        coreApplicationContext.refresh();
+
+        dataPrepper = coreApplicationContext.getBean(DataPrepper.class);
+        dataPrepper.registerShutdownHandler(new ContextShutdownListener());
+
+        LOG.trace("Data Prepper context is fully refreshed.");
+    }
+
+    /**
+     * Override this method to modify the public {@link AnnotationConfigApplicationContext} before
+     * the {@link AbstractContextManager} calls {@link AnnotationConfigApplicationContext#refresh()}
+     *
+     * @param publicApplicationContext The public application context
+     */
+    protected void preRefreshPublicApplicationContext(final AnnotationConfigApplicationContext publicApplicationContext) {
+
+    }
+
+    /**
+     * Override this method to modify the core {@link AnnotationConfigApplicationContext} before
+     * the {@link AbstractContextManager} calls {@link AnnotationConfigApplicationContext#refresh()}
+     *
+     * @param coreApplicationContext The core application context
+     */
+    protected void preRefreshCoreApplicationContext(final AnnotationConfigApplicationContext coreApplicationContext) {
+
+    }
+
+    public void shutdown() {
+        coreApplicationContext.close();
+        publicApplicationContext.close();
+    }
+
+    private class ContextShutdownListener implements DataPrepperShutdownListener {
+        @Override
+        public void handleShutdown() {
+            shutdown();
+        }
+    }
+}

--- a/data-prepper-core/src/main/java/org/opensearch/dataprepper/DataPrepper.java
+++ b/data-prepper-core/src/main/java/org/opensearch/dataprepper/DataPrepper.java
@@ -9,6 +9,8 @@ import org.opensearch.dataprepper.model.plugin.PluginFactory;
 import org.opensearch.dataprepper.parser.PipelineParser;
 import org.opensearch.dataprepper.peerforwarder.server.PeerForwarderServer;
 import org.opensearch.dataprepper.pipeline.Pipeline;
+import org.opensearch.dataprepper.pipeline.PipelineObserver;
+import org.opensearch.dataprepper.pipeline.PipelinesProvider;
 import org.opensearch.dataprepper.pipeline.server.DataPrepperServer;
 import io.micrometer.core.instrument.util.StringUtils;
 import org.slf4j.Logger;
@@ -25,11 +27,11 @@ import java.util.Iterator;
 /**
  * DataPrepper is the entry point into the execution engine. The instance can be used to trigger execution via
  * {@link #execute()} of the {@link Pipeline} with default configuration or {@link #execute()} to
- * provide custom configuration file. Also, the same instance reference can be further used to {@link #shutdown()} the
+ * provide custom configuration file. Also, the same instance reference can be further used to {@link #shutdownPipelines()} the
  * execution.
  */
 @Named
-public class DataPrepper {
+public class DataPrepper implements PipelinesProvider {
     private static final Logger LOG = LoggerFactory.getLogger(DataPrepper.class);
     private static final String DATAPREPPER_SERVICE_NAME = "DATAPREPPER_SERVICE_NAME";
     private static final String DEFAULT_SERVICE_NAME = "dataprepper";
@@ -37,12 +39,14 @@ public class DataPrepper {
 
     private final PluginFactory pluginFactory;
     private final PeerForwarderServer peerForwarderServer;
-    private Map<String, Pipeline> transformationPipelines;
+    private final PipelinesObserver pipelinesObserver;
+    private final Map<String, Pipeline> transformationPipelines;
 
     // TODO: Remove DataPrepperServer dependency on DataPrepper
     @Inject
     @Lazy
     private DataPrepperServer dataPrepperServer;
+    private DataPrepperShutdownListener shutdownListener;
 
     /**
      * returns serviceName if exists or default serviceName
@@ -66,6 +70,7 @@ public class DataPrepper {
             throw new RuntimeException("No valid pipeline is available for execution, exiting");
         }
         this.peerForwarderServer = peerForwarderServer;
+        pipelinesObserver = new PipelinesObserver();
     }
 
     /**
@@ -95,9 +100,12 @@ public class DataPrepper {
         }
         if (waitingPipelineNames.size() > 0) {
             LOG.info("One or more Pipelines are not ready even after {} retries. Shutting down pipelines", numRetries);
-            shutdown();
+            shutdownPipelines();
             throw new RuntimeException("Failed to start pipelines");
         }
+        transformationPipelines.forEach((name, pipeline) -> {
+            pipeline.addShutdownObserver(pipelinesObserver);
+        });
         transformationPipelines.forEach((name, pipeline) -> {
             pipeline.execute();
         });
@@ -105,10 +113,19 @@ public class DataPrepper {
         return true;
     }
 
+    public void shutdown() {
+        shutdownPipelines();
+        shutdownServers();
+    }
+
     /**
      * Triggers the shutdown of all configured valid pipelines.
      */
-    public void shutdown() {
+    public void shutdownPipelines() {
+        transformationPipelines.forEach((name, pipeline) -> {
+            pipeline.removeShutdownObserver(pipelinesObserver);
+        });
+
         for (final Pipeline pipeline : transformationPipelines.values()) {
             LOG.info("Shutting down pipeline: {}", pipeline.getName());
             pipeline.shutdown();
@@ -121,6 +138,9 @@ public class DataPrepper {
     public void shutdownServers() {
         dataPrepperServer.stop();
         peerForwarderServer.stop();
+        if(shutdownListener != null) {
+            shutdownListener.handleShutdown();
+        }
     }
 
     /**
@@ -128,7 +148,7 @@ public class DataPrepper {
      *
      * @param pipeline name of the pipeline
      */
-    public void shutdown(final String pipeline) {
+    public void shutdownPipelines(final String pipeline) {
         if (transformationPipelines.containsKey(pipeline)) {
             transformationPipelines.get(pipeline).shutdown();
         }
@@ -139,5 +159,22 @@ public class DataPrepper {
 
     public Map<String, Pipeline> getTransformationPipelines() {
         return transformationPipelines;
+    }
+
+    public void registerShutdownHandler(final DataPrepperShutdownListener shutdownListener) {
+        this.shutdownListener = shutdownListener;
+    }
+
+    private class PipelinesObserver implements PipelineObserver {
+        @Override
+        public void shutdown(final Pipeline pipeline) {
+            pipeline.removeShutdownObserver(pipelinesObserver);
+            transformationPipelines.remove(pipeline.getName());
+
+            LOG.info("Pipeline has shutdown. '{}'. There are {} pipelines remaining.", pipeline.getName(), transformationPipelines.size());
+            if(transformationPipelines.size() == 0) {
+                DataPrepper.this.shutdown();
+            }
+        }
     }
 }

--- a/data-prepper-core/src/main/java/org/opensearch/dataprepper/DataPrepperShutdownListener.java
+++ b/data-prepper-core/src/main/java/org/opensearch/dataprepper/DataPrepperShutdownListener.java
@@ -1,0 +1,10 @@
+/*
+ * Copyright OpenSearch Contributors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package org.opensearch.dataprepper;
+
+interface DataPrepperShutdownListener {
+    void handleShutdown();
+}

--- a/data-prepper-core/src/main/java/org/opensearch/dataprepper/pipeline/Pipeline.java
+++ b/data-prepper-core/src/main/java/org/opensearch/dataprepper/pipeline/Pipeline.java
@@ -6,22 +6,22 @@
 package org.opensearch.dataprepper.pipeline;
 
 import com.google.common.base.Preconditions;
+import org.opensearch.dataprepper.acknowledgements.InactiveAcknowledgementSetManager;
+import org.opensearch.dataprepper.model.acknowledgements.AcknowledgementSetManager;
 import org.opensearch.dataprepper.model.buffer.Buffer;
+import org.opensearch.dataprepper.model.event.EventFactory;
 import org.opensearch.dataprepper.model.processor.Processor;
 import org.opensearch.dataprepper.model.record.Record;
 import org.opensearch.dataprepper.model.sink.Sink;
-import org.opensearch.dataprepper.model.source.coordinator.UsesSourceCoordination;
 import org.opensearch.dataprepper.model.source.Source;
 import org.opensearch.dataprepper.model.source.coordinator.SourceCoordinator;
+import org.opensearch.dataprepper.model.source.coordinator.UsesSourceCoordination;
 import org.opensearch.dataprepper.parser.DataFlowComponent;
 import org.opensearch.dataprepper.pipeline.common.PipelineThreadFactory;
 import org.opensearch.dataprepper.pipeline.common.PipelineThreadPoolExecutor;
 import org.opensearch.dataprepper.pipeline.router.Router;
 import org.opensearch.dataprepper.pipeline.router.RouterCopyRecordStrategy;
 import org.opensearch.dataprepper.pipeline.router.RouterGetRecordStrategy;
-import org.opensearch.dataprepper.model.event.EventFactory;
-import org.opensearch.dataprepper.model.acknowledgements.AcknowledgementSetManager;
-import org.opensearch.dataprepper.acknowledgements.InactiveAcknowledgementSetManager;
 import org.opensearch.dataprepper.sourcecoordination.SourceCoordinatorFactory;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
@@ -30,6 +30,8 @@ import javax.annotation.Nonnull;
 import java.time.Duration;
 import java.util.ArrayList;
 import java.util.Collection;
+import java.util.Collections;
+import java.util.LinkedList;
 import java.util.List;
 import java.util.Objects;
 import java.util.concurrent.ExecutorService;
@@ -65,6 +67,7 @@ public class Pipeline {
     private final ExecutorService sinkExecutorService;
     private final EventFactory eventFactory;
     private final AcknowledgementSetManager acknowledgementSetManager;
+    private final List<PipelineObserver> observers = Collections.synchronizedList(new LinkedList<>());
 
     /**
      * Constructs a {@link Pipeline} object with provided {@link Source}, {@link #name}, {@link Collection} of
@@ -250,28 +253,40 @@ public class Pipeline {
                     "proceeding with termination of process workers", name);
         }
 
-        shutdownExecutorService(processorExecutorService, processorShutdownTimeout.toMillis());
+        shutdownExecutorService(processorExecutorService, processorShutdownTimeout.toMillis(), "processor");
 
         processorSets.forEach(processorSet -> processorSet.forEach(Processor::shutdown));
         sinks.stream()
                 .map(DataFlowComponent::getComponent)
                 .forEach(Sink::shutdown);
 
-        shutdownExecutorService(sinkExecutorService, sinkShutdownTimeout.toMillis());
+        shutdownExecutorService(sinkExecutorService, sinkShutdownTimeout.toMillis(), "sink");
+
+        LOG.info("Pipeline [{}] - Pipeline fully shutdown.", name);
+
+        observers.forEach(observer -> observer.shutdown(this));
     }
 
-    private void shutdownExecutorService(final ExecutorService executorService, final long timeoutForTerminationInMillis) {
-        LOG.info("Pipeline [{}] - Shutting down process workers", name);
+    public void addShutdownObserver(final PipelineObserver pipelineObserver) {
+        observers.add(pipelineObserver);
+    }
+
+    public void removeShutdownObserver(final PipelineObserver pipelineObserver) {
+        observers.remove(pipelineObserver);
+    }
+
+    private void shutdownExecutorService(final ExecutorService executorService, final long timeoutForTerminationInMillis, final String workerName) {
+        LOG.info("Pipeline [{}] - Shutting down {} process workers.", name, workerName);
 
         executorService.shutdown();
         try {
             if (!executorService.awaitTermination(timeoutForTerminationInMillis, TimeUnit.MILLISECONDS)) {
-                LOG.warn("Pipeline [{}] - Workers did not terminate in time, forcing termination", name);
+                LOG.warn("Pipeline [{}] - Workers did not terminate in time, forcing termination of {} workers.", name, workerName);
                 executorService.shutdownNow();
             }
         } catch (InterruptedException ex) {
             LOG.info("Pipeline [{}] - Encountered interruption terminating the pipeline execution, " +
-                    "Attempting to force the termination", name);
+                    "Attempting to force the termination of {} workers.", name, workerName);
             executorService.shutdownNow();
         }
     }

--- a/data-prepper-core/src/main/java/org/opensearch/dataprepper/pipeline/PipelineObserver.java
+++ b/data-prepper-core/src/main/java/org/opensearch/dataprepper/pipeline/PipelineObserver.java
@@ -1,0 +1,13 @@
+/*
+ * Copyright OpenSearch Contributors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package org.opensearch.dataprepper.pipeline;
+
+/**
+ * Allows an observer to observe major changes in pipelines.
+ */
+public interface PipelineObserver {
+    void shutdown(Pipeline pipeline);
+}

--- a/data-prepper-core/src/main/java/org/opensearch/dataprepper/pipeline/PipelinesProvider.java
+++ b/data-prepper-core/src/main/java/org/opensearch/dataprepper/pipeline/PipelinesProvider.java
@@ -1,0 +1,15 @@
+/*
+ * Copyright OpenSearch Contributors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package org.opensearch.dataprepper.pipeline;
+
+import java.util.Map;
+
+/**
+ * Interface for a provider of available Data Prepper Pipelines.
+ */
+public interface PipelinesProvider {
+    Map<String, Pipeline> getTransformationPipelines();
+}

--- a/data-prepper-core/src/main/java/org/opensearch/dataprepper/pipeline/server/ListPipelinesHandler.java
+++ b/data-prepper-core/src/main/java/org/opensearch/dataprepper/pipeline/server/ListPipelinesHandler.java
@@ -5,10 +5,10 @@
 
 package org.opensearch.dataprepper.pipeline.server;
 
-import org.opensearch.dataprepper.DataPrepper;
 import com.fasterxml.jackson.databind.ObjectMapper;
 import com.sun.net.httpserver.HttpExchange;
 import com.sun.net.httpserver.HttpHandler;
+import org.opensearch.dataprepper.pipeline.PipelinesProvider;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
@@ -24,12 +24,12 @@ import java.util.stream.Collectors;
  */
 public class ListPipelinesHandler implements HttpHandler {
 
-    private final DataPrepper dataPrepper;
+    private final PipelinesProvider pipelinesProvider;
     private static final ObjectMapper OBJECT_MAPPER = new ObjectMapper();
     private final Logger LOG = LoggerFactory.getLogger(ListPipelinesHandler.class);
 
-    public ListPipelinesHandler(final DataPrepper dataPrepper) {
-        this.dataPrepper = dataPrepper;
+    public ListPipelinesHandler(final PipelinesProvider pipelinesProvider) {
+        this.pipelinesProvider = pipelinesProvider;
     }
 
     private static class PipelineListing {
@@ -50,7 +50,7 @@ public class ListPipelinesHandler implements HttpHandler {
         }
 
         try {
-            final List<PipelineListing> pipelines = dataPrepper.getTransformationPipelines()
+            final List<PipelineListing> pipelines = pipelinesProvider.getTransformationPipelines()
                     .keySet()
                     .stream()
                     .map(PipelineListing::new)

--- a/data-prepper-core/src/main/java/org/opensearch/dataprepper/pipeline/server/ShutdownHandler.java
+++ b/data-prepper-core/src/main/java/org/opensearch/dataprepper/pipeline/server/ShutdownHandler.java
@@ -36,7 +36,7 @@ public class ShutdownHandler implements HttpHandler {
         }
 
         try {
-            dataPrepper.shutdown();
+            dataPrepper.shutdownPipelines();
             exchange.sendResponseHeaders(HttpURLConnection.HTTP_OK, 0);
         } catch (Exception e) {
             LOG.error("Caught exception shutting down data prepper", e);

--- a/data-prepper-core/src/main/java/org/opensearch/dataprepper/pipeline/server/config/DataPrepperServerConfiguration.java
+++ b/data-prepper-core/src/main/java/org/opensearch/dataprepper/pipeline/server/config/DataPrepperServerConfiguration.java
@@ -9,6 +9,7 @@ import org.opensearch.dataprepper.DataPrepper;
 import org.opensearch.dataprepper.model.configuration.PluginModel;
 import org.opensearch.dataprepper.model.configuration.PluginSetting;
 import org.opensearch.dataprepper.model.plugin.PluginFactory;
+import org.opensearch.dataprepper.pipeline.PipelinesProvider;
 import org.opensearch.dataprepper.pipeline.server.DataPrepperCoreAuthenticationProvider;
 import org.opensearch.dataprepper.pipeline.server.ListPipelinesHandler;
 import org.opensearch.dataprepper.pipeline.server.ShutdownHandler;
@@ -65,8 +66,8 @@ public class DataPrepperServerConfiguration {
     }
 
     @Bean
-    public ListPipelinesHandler listPipelinesHandler(final DataPrepper dataPrepper) {
-        return new ListPipelinesHandler(dataPrepper);
+    public ListPipelinesHandler listPipelinesHandler(final PipelinesProvider pipelinesProvider) {
+        return new ListPipelinesHandler(pipelinesProvider);
     }
 
     @Bean

--- a/data-prepper-core/src/test/java/org/opensearch/dataprepper/AbstractContextManagerTest.java
+++ b/data-prepper-core/src/test/java/org/opensearch/dataprepper/AbstractContextManagerTest.java
@@ -1,0 +1,108 @@
+/*
+ * Copyright OpenSearch Contributors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package org.opensearch.dataprepper;
+
+import com.google.common.collect.Lists;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+import org.mockito.ArgumentCaptor;
+import org.mockito.InOrder;
+import org.mockito.MockedConstruction;
+import org.springframework.context.annotation.AnnotationConfigApplicationContext;
+
+import java.util.ArrayList;
+import java.util.List;
+
+import static org.hamcrest.CoreMatchers.notNullValue;
+import static org.hamcrest.MatcherAssert.assertThat;
+import static org.mockito.ArgumentMatchers.any;
+import static org.mockito.Mockito.inOrder;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.mockConstruction;
+import static org.mockito.Mockito.never;
+import static org.mockito.Mockito.spy;
+import static org.mockito.Mockito.verify;
+import static org.mockito.Mockito.when;
+
+class AbstractContextManagerTest {
+    private List<AnnotationConfigApplicationContext> applicationContexts;
+    private DataPrepper dataPrepper;
+
+    @BeforeEach
+    void setUp() {
+        applicationContexts = new ArrayList<>();
+
+        dataPrepper = mock(DataPrepper.class);
+    }
+
+    private AbstractContextManager createObjectUnderTest() {
+        try(final MockedConstruction<AnnotationConfigApplicationContext> mockConstruction = mockConstruction(AnnotationConfigApplicationContext.class)) {
+            final AbstractContextManager objectUnderTest = spy(AbstractContextManager.class);
+            applicationContexts.addAll(mockConstruction.constructed());
+            return objectUnderTest;
+        }
+    }
+
+    @Test
+    void getDataPrepperBean_calls_pre_refresh_hooks() {
+        final AbstractContextManager objectUnderTest = createObjectUnderTest();
+
+        assertThat(applicationContexts, notNullValue());
+
+        final AnnotationConfigApplicationContext coreContext = applicationContexts.get(1);
+        when(coreContext.getBean(DataPrepper.class)).thenReturn(dataPrepper);
+
+        verify(objectUnderTest, never()).preRefreshPublicApplicationContext(any());
+        verify(objectUnderTest, never()).preRefreshCoreApplicationContext(any());
+
+        objectUnderTest.getDataPrepperBean();
+
+        verify(objectUnderTest).preRefreshPublicApplicationContext(applicationContexts.get(0));
+        verify(objectUnderTest).preRefreshCoreApplicationContext(applicationContexts.get(1));
+    }
+
+    @Test
+    void getDataPrepperBean_calls_refresh_once_across_multiple_calls() {
+        final AbstractContextManager objectUnderTest = createObjectUnderTest();
+
+        assertThat(applicationContexts, notNullValue());
+
+        final AnnotationConfigApplicationContext coreContext = applicationContexts.get(1);
+        when(coreContext.getBean(DataPrepper.class)).thenReturn(dataPrepper);
+
+        for (int i = 0; i < 3; i++) {
+            objectUnderTest.getDataPrepperBean();
+        }
+
+        verify(applicationContexts.get(0)).refresh();
+        verify(applicationContexts.get(1)).refresh();
+    }
+
+    @Test
+    void getDataPrepperBean_closes_application_contexts_when_DataPrepper_shutsdown() {
+        final AbstractContextManager objectUnderTest = createObjectUnderTest();
+
+        assertThat(applicationContexts, notNullValue());
+
+        final AnnotationConfigApplicationContext coreContext = applicationContexts.get(1);
+        when(coreContext.getBean(DataPrepper.class)).thenReturn(dataPrepper);
+
+        objectUnderTest.getDataPrepperBean();
+
+        ArgumentCaptor<DataPrepperShutdownListener> dataPrepperShutdownListenerArgumentCaptor = ArgumentCaptor.forClass(DataPrepperShutdownListener.class);
+        verify(dataPrepper).registerShutdownHandler(dataPrepperShutdownListenerArgumentCaptor.capture());
+
+        final DataPrepperShutdownListener shutdownListener = dataPrepperShutdownListenerArgumentCaptor.getValue();
+
+        shutdownListener.handleShutdown();
+
+        final InOrder inOrder = inOrder(applicationContexts.toArray());
+        final List<AnnotationConfigApplicationContext> reversedContexts = Lists.reverse(applicationContexts);
+        for (AnnotationConfigApplicationContext applicationContext : reversedContexts) {
+            inOrder.verify(applicationContext).close();
+        }
+    }
+}

--- a/data-prepper-core/src/test/java/org/opensearch/dataprepper/DataPrepperTests.java
+++ b/data-prepper-core/src/test/java/org/opensearch/dataprepper/DataPrepperTests.java
@@ -5,70 +5,72 @@
 
 package org.opensearch.dataprepper;
 
+import org.hamcrest.Matchers;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.ExtendWith;
+import org.mockito.ArgumentCaptor;
+import org.mockito.Mock;
+import org.mockito.junit.jupiter.MockitoExtension;
 import org.opensearch.dataprepper.model.plugin.PluginFactory;
 import org.opensearch.dataprepper.parser.PipelineParser;
 import org.opensearch.dataprepper.peerforwarder.server.PeerForwarderServer;
 import org.opensearch.dataprepper.pipeline.Pipeline;
+import org.opensearch.dataprepper.pipeline.PipelineObserver;
 import org.opensearch.dataprepper.pipeline.server.DataPrepperServer;
-import org.hamcrest.Matchers;
-import org.junit.jupiter.api.BeforeAll;
-import org.junit.jupiter.api.BeforeEach;
-import org.junit.jupiter.api.Test;
-import org.junit.jupiter.api.extension.ExtendWith;
-import org.mockito.InjectMocks;
-import org.mockito.Mock;
-import org.mockito.junit.jupiter.MockitoExtension;
-import static org.mockito.Mockito.lenient;
 
 import java.lang.reflect.Field;
 import java.util.HashMap;
 import java.util.Map;
 
+import static org.hamcrest.CoreMatchers.equalTo;
+import static org.hamcrest.CoreMatchers.notNullValue;
 import static org.hamcrest.MatcherAssert.assertThat;
 import static org.junit.jupiter.api.Assertions.assertThrows;
+import static org.mockito.Mockito.lenient;
 import static org.mockito.Mockito.mock;
 import static org.mockito.Mockito.verify;
-import static org.mockito.Mockito.when;
 
 @ExtendWith(MockitoExtension.class)
 public class DataPrepperTests {
-    private static final PipelineParser pipelineParser = mock(PipelineParser.class);
-    private static final Pipeline pipeline = mock(Pipeline.class);
-    private static Map<String, Pipeline> parseConfigurationFixture;
-
+    private Map<String, Pipeline> parseConfigurationFixture;
+    @Mock
+    private PipelineParser pipelineParser;
+    @Mock
+    private Pipeline pipeline;
     @Mock
     private PluginFactory pluginFactory;
     @Mock
     private DataPrepperServer dataPrepperServer;
     @Mock
     private PeerForwarderServer peerForwarderServer;
-    @InjectMocks
-    private DataPrepper dataPrepper;
 
-    @BeforeAll
-    public static void beforeAll() {
+    @BeforeEach
+    public void before() {
         parseConfigurationFixture = new HashMap<>();
         parseConfigurationFixture.put("testKey", pipeline);
-        when(pipeline.isReady()).thenReturn(true);
+        lenient().when(pipeline.getName()).thenReturn("testKey");
+        lenient().when(pipeline.isReady()).thenReturn(true);
 
-        when(pipelineParser.parseConfiguration())
+        lenient().when(pipelineParser.parseConfiguration())
                 .thenReturn(parseConfigurationFixture);
     }
 
-    @BeforeEach
-    public void before() throws NoSuchFieldException, IllegalAccessException {
-        // Use reflection to set dataPrepper.dataPrepperServer because @InjectMock will not use field injection.
+    private DataPrepper createObjectUnderTest() throws NoSuchFieldException, IllegalAccessException {
+        final DataPrepper dataPrepper = new DataPrepper(pipelineParser, pluginFactory, peerForwarderServer);
         final Field dataPrepperServerField = dataPrepper.getClass().getDeclaredField("dataPrepperServer");
         dataPrepperServerField.setAccessible(true);
         dataPrepperServerField.set(dataPrepper, dataPrepperServer);
         dataPrepperServerField.setAccessible(false);
+
+        return dataPrepper;
     }
 
     @Test
-    public void testGivenValidInputThenInstanceCreation() {
+    public void testGivenValidInputThenInstanceCreation() throws NoSuchFieldException, IllegalAccessException {
         assertThat(
                 "Given injected with valid beans a DataPrepper bean should be available",
-                dataPrepper,
+                createObjectUnderTest(),
                 Matchers.is(Matchers.notNullValue()));
     }
 
@@ -83,18 +85,18 @@ public class DataPrepperTests {
     }
 
     @Test
-    public void testGivenInstantiatedWithPluginFactoryWhenGetPluginFactoryCalledThenReturnSamePluginFactory() {
-        assertThat(dataPrepper.getPluginFactory(), Matchers.is(pluginFactory));
+    public void testGivenInstantiatedWithPluginFactoryWhenGetPluginFactoryCalledThenReturnSamePluginFactory() throws NoSuchFieldException, IllegalAccessException {
+        assertThat(createObjectUnderTest().getPluginFactory(), Matchers.is(pluginFactory));
     }
 
     @Test
-    public void testGivenValidPipelineParserThenReturnResultOfParseConfiguration() {
-        assertThat(dataPrepper.getTransformationPipelines(), Matchers.is(parseConfigurationFixture));
+    public void testGivenValidPipelineParserThenReturnResultOfParseConfiguration() throws NoSuchFieldException, IllegalAccessException {
+        assertThat(createObjectUnderTest().getTransformationPipelines(), Matchers.is(parseConfigurationFixture));
     }
 
     @Test
-    public void testGivenValidPipelineParserWhenExecuteThenAllPipelinesExecuteAndServerStartAndReturnTrue() {
-        assertThat(dataPrepper.execute(), Matchers.is(true));
+    public void testGivenValidPipelineParserWhenExecuteThenAllPipelinesExecuteAndServerStartAndReturnTrue() throws NoSuchFieldException, IllegalAccessException {
+        assertThat(createObjectUnderTest().execute(), Matchers.is(true));
 
         verify(pipeline).execute();
         verify(dataPrepperServer).start();
@@ -102,29 +104,29 @@ public class DataPrepperTests {
     }
 
     @Test
-    public void testDataPrepperShutdown() {
-        dataPrepper.shutdown();
+    public void testDataPrepperShutdown() throws NoSuchFieldException, IllegalAccessException {
+        createObjectUnderTest().shutdownPipelines();
         verify(pipeline).shutdown();
     }
 
     @Test
-    public void testDataPrepperShutdownPipeline() {
+    public void testDataPrepperShutdownPipeline() throws NoSuchFieldException, IllegalAccessException {
         final Pipeline randomPipeline = mock(Pipeline.class);
         lenient().when(randomPipeline.isReady()).thenReturn(true);
         parseConfigurationFixture.put("Random Pipeline", randomPipeline);
-        dataPrepper.shutdown("Random Pipeline");
+        createObjectUnderTest().shutdownPipelines("Random Pipeline");
 
         verify(randomPipeline).shutdown();
     }
 
     @Test
-    public void testDataPrepperShutdownNonExistentPipelineWithoutException() {
-        dataPrepper.shutdown("Missing Pipeline");
+    public void testDataPrepperShutdownNonExistentPipelineWithoutException() throws NoSuchFieldException, IllegalAccessException {
+        createObjectUnderTest().shutdownPipelines("Missing Pipeline");
     }
 
     @Test
-    public void testShutdownServers() {
-        dataPrepper.shutdownServers();
+    public void testShutdownServers() throws NoSuchFieldException, IllegalAccessException {
+        createObjectUnderTest().shutdownServers();
 
         verify(dataPrepperServer).stop();
         verify(peerForwarderServer).stop();
@@ -135,5 +137,23 @@ public class DataPrepperTests {
         final String actual = DataPrepper.getServiceNameForMetrics();
 
         assertThat(actual, Matchers.is("dataprepper"));
+    }
+
+    @Test
+    void dataPrepper_shuts_down_when_last_pipeline_is_shutdown() throws NoSuchFieldException, IllegalAccessException {
+        final DataPrepper objectUnderTest = createObjectUnderTest();
+        objectUnderTest.execute();
+        final ArgumentCaptor<PipelineObserver> pipelineObserverArgumentCaptor = ArgumentCaptor.forClass(PipelineObserver.class);
+        verify(pipeline).addShutdownObserver(pipelineObserverArgumentCaptor.capture());
+
+        final PipelineObserver pipelineObserver = pipelineObserverArgumentCaptor.getValue();
+        pipelineObserver.shutdown(pipeline);
+
+        verify(dataPrepperServer).stop();
+        verify(peerForwarderServer).stop();
+
+        final Map<String, Pipeline> transformationPipelines = objectUnderTest.getTransformationPipelines();
+        assertThat(transformationPipelines, notNullValue());
+        assertThat(transformationPipelines.size(), equalTo(0));
     }
 }

--- a/data-prepper-core/src/test/java/org/opensearch/dataprepper/pipeline/PipelineTests.java
+++ b/data-prepper-core/src/test/java/org/opensearch/dataprepper/pipeline/PipelineTests.java
@@ -482,4 +482,49 @@ class PipelineTests {
             }
         }
     }
+
+    @Test
+    void shutdown_calls_added_PipelineObservers() {
+        final Source<Record<String>> testSource = new TestSource();
+        final DataFlowComponent<Sink> sinkDataFlowComponent = mock(DataFlowComponent.class);
+        final TestSink testSink = new TestSink();
+        when(sinkDataFlowComponent.getComponent()).thenReturn(testSink);
+        testPipeline = new Pipeline(TEST_PIPELINE_NAME, testSource, new BlockingBuffer(TEST_PIPELINE_NAME),
+                Collections.emptyList(), Collections.singletonList(sinkDataFlowComponent), router,
+                eventFactory, acknowledgementSetManager, sourceCoordinatorFactory, TEST_PROCESSOR_THREADS, TEST_READ_BATCH_TIMEOUT,
+                processorShutdownTimeout, sinkShutdownTimeout, peerForwarderDrainTimeout);
+
+        PipelineObserver pipelineObserver = mock(PipelineObserver.class);
+        testPipeline.addShutdownObserver(pipelineObserver);
+
+        testPipeline.execute();
+
+        verifyNoInteractions(pipelineObserver);
+
+        testPipeline.shutdown();
+
+        verify(pipelineObserver).shutdown(testPipeline);
+    }
+
+    @Test
+    void shutdown_does_not_call_removed_PipelineObservers() {
+        final Source<Record<String>> testSource = new TestSource();
+        final DataFlowComponent<Sink> sinkDataFlowComponent = mock(DataFlowComponent.class);
+        final TestSink testSink = new TestSink();
+        when(sinkDataFlowComponent.getComponent()).thenReturn(testSink);
+        testPipeline = new Pipeline(TEST_PIPELINE_NAME, testSource, new BlockingBuffer(TEST_PIPELINE_NAME),
+                Collections.emptyList(), Collections.singletonList(sinkDataFlowComponent), router,
+                eventFactory, acknowledgementSetManager, sourceCoordinatorFactory, TEST_PROCESSOR_THREADS, TEST_READ_BATCH_TIMEOUT,
+                processorShutdownTimeout, sinkShutdownTimeout, peerForwarderDrainTimeout);
+
+        PipelineObserver pipelineObserver = mock(PipelineObserver.class);
+        testPipeline.addShutdownObserver(pipelineObserver);
+
+        testPipeline.execute();
+
+        testPipeline.removeShutdownObserver(pipelineObserver);
+
+        testPipeline.shutdown();
+        verifyNoInteractions(pipelineObserver);
+    }
 }

--- a/data-prepper-core/src/test/java/org/opensearch/dataprepper/pipeline/server/ListPipelinesHandlerTest.java
+++ b/data-prepper-core/src/test/java/org/opensearch/dataprepper/pipeline/server/ListPipelinesHandlerTest.java
@@ -5,16 +5,16 @@
 
 package org.opensearch.dataprepper.pipeline.server;
 
+import com.sun.net.httpserver.Headers;
+import com.sun.net.httpserver.HttpExchange;
 import org.junit.jupiter.api.BeforeEach;
 import org.junit.jupiter.api.extension.ExtendWith;
 import org.junit.jupiter.params.ParameterizedTest;
 import org.junit.jupiter.params.provider.ValueSource;
 import org.mockito.Mock;
 import org.mockito.junit.jupiter.MockitoExtension;
-import org.opensearch.dataprepper.DataPrepper;
 import org.opensearch.dataprepper.pipeline.Pipeline;
-import com.sun.net.httpserver.Headers;
-import com.sun.net.httpserver.HttpExchange;
+import org.opensearch.dataprepper.pipeline.PipelinesProvider;
 
 import javax.ws.rs.HttpMethod;
 import java.io.IOException;
@@ -33,6 +33,8 @@ import static org.mockito.Mockito.when;
 @ExtendWith(MockitoExtension.class)
 class ListPipelinesHandlerTest {
     @Mock
+    private PipelinesProvider pipelinesProvider;
+    @Mock
     private HttpExchange httpExchange;
 
     @Mock
@@ -44,21 +46,24 @@ class ListPipelinesHandlerTest {
                 .thenReturn(outputStream);
     }
 
+    private ListPipelinesHandler createObjectUnderTest() {
+        return new ListPipelinesHandler(pipelinesProvider);
+    }
+
     @ParameterizedTest
     @ValueSource(strings = { HttpMethod.GET, HttpMethod.POST })
     public void testGivenNoPipelinesThenResponseWritten(String httpMethod) throws IOException {
-        final DataPrepper dataPrepper = mock(DataPrepper.class);
         final Headers headers = mock(Headers.class);
         final Map<String, Pipeline> transformationPipelines = new HashMap<>();
 
-        when(dataPrepper.getTransformationPipelines())
+        when(pipelinesProvider.getTransformationPipelines())
                 .thenReturn(transformationPipelines);
         when(httpExchange.getResponseHeaders())
                 .thenReturn(headers);
         when(httpExchange.getRequestMethod())
                 .thenReturn(httpMethod);
 
-        final ListPipelinesHandler handler = new ListPipelinesHandler(dataPrepper);
+        final ListPipelinesHandler handler = createObjectUnderTest();
 
         handler.handle(httpExchange);
 
@@ -75,7 +80,6 @@ class ListPipelinesHandlerTest {
     @ParameterizedTest
     @ValueSource(strings = { HttpMethod.GET, HttpMethod.POST })
     public void testGivenPipelinesThenResponseWritten(String httpMethod) throws IOException {
-        final DataPrepper dataPrepper = mock(DataPrepper.class);
         final Headers headers = mock(Headers.class);
         final Pipeline pipeline = mock(Pipeline.class);
         final Map<String, Pipeline> transformationPipelines = new HashMap<>();
@@ -83,14 +87,14 @@ class ListPipelinesHandlerTest {
         transformationPipelines.put("Pipeline B", pipeline);
         transformationPipelines.put("Pipeline C", pipeline);
 
-        when(dataPrepper.getTransformationPipelines())
+        when(pipelinesProvider.getTransformationPipelines())
                 .thenReturn(transformationPipelines);
         when(httpExchange.getResponseHeaders())
                 .thenReturn(headers);
         when(httpExchange.getRequestMethod())
                 .thenReturn(httpMethod);
 
-        final ListPipelinesHandler handler = new ListPipelinesHandler(dataPrepper);
+        final ListPipelinesHandler handler = createObjectUnderTest();
 
         handler.handle(httpExchange);
 
@@ -107,8 +111,7 @@ class ListPipelinesHandlerTest {
     @ParameterizedTest
     @ValueSource(strings = { HttpMethod.DELETE, HttpMethod.PATCH, HttpMethod.PUT })
     public void testGivenProhibitedHttpMethodThenErrorResponseWritten(String httpMethod) throws IOException {
-        final DataPrepper dataPrepper = mock(DataPrepper.class);
-        final ListPipelinesHandler handler = new ListPipelinesHandler(dataPrepper);
+        final ListPipelinesHandler handler = createObjectUnderTest();
 
         when(httpExchange.getRequestMethod())
                 .thenReturn(httpMethod);
@@ -127,7 +130,8 @@ class ListPipelinesHandlerTest {
         when(httpExchange.getRequestMethod())
                 .thenReturn(httpMethod);
 
-        final ListPipelinesHandler handler = new ListPipelinesHandler(null);
+        pipelinesProvider = null;
+        final ListPipelinesHandler handler = createObjectUnderTest();
         handler.handle(httpExchange);
 
         verify(httpExchange)

--- a/data-prepper-core/src/test/java/org/opensearch/dataprepper/pipeline/server/ShutdownHandlerTest.java
+++ b/data-prepper-core/src/test/java/org/opensearch/dataprepper/pipeline/server/ShutdownHandlerTest.java
@@ -55,7 +55,7 @@ public class ShutdownHandlerTest {
         shutdownHandler.handle(exchange);
 
         verify(dataPrepper, times(1))
-                .shutdown();
+                .shutdownPipelines();
         verify(exchange, times(1))
                 .sendResponseHeaders(eq(HttpURLConnection.HTTP_OK), eq(0L));
         verify(responseBody, times(1))
@@ -82,7 +82,7 @@ public class ShutdownHandlerTest {
     public void testHandleException() throws IOException {
         when(exchange.getRequestMethod())
                 .thenReturn(HttpMethod.POST);
-        doThrow(RuntimeException.class).when(dataPrepper).shutdown();
+        doThrow(RuntimeException.class).when(dataPrepper).shutdownPipelines();
 
         shutdownHandler.handle(exchange);
 

--- a/data-prepper-core/src/test/java/org/opensearch/dataprepper/pipeline/server/config/DataPrepperServerConfigurationTest.java
+++ b/data-prepper-core/src/test/java/org/opensearch/dataprepper/pipeline/server/config/DataPrepperServerConfigurationTest.java
@@ -9,6 +9,7 @@ import org.opensearch.dataprepper.DataPrepper;
 import org.opensearch.dataprepper.model.configuration.PluginModel;
 import org.opensearch.dataprepper.model.configuration.PluginSetting;
 import org.opensearch.dataprepper.model.plugin.PluginFactory;
+import org.opensearch.dataprepper.pipeline.PipelinesProvider;
 import org.opensearch.dataprepper.pipeline.server.DataPrepperCoreAuthenticationProvider;
 import org.opensearch.dataprepper.pipeline.server.ListPipelinesHandler;
 import org.opensearch.dataprepper.pipeline.server.ShutdownHandler;
@@ -100,18 +101,18 @@ class DataPrepperServerConfigurationTest {
 
     @Test
     public void testGivenValidInputWithNoAuthenticatorThenServerListContextCreated() {
-        final DataPrepper dataPrepper = mock(DataPrepper.class);
+        final PipelinesProvider pipelinesProvider = mock(PipelinesProvider.class);
 
-        final ListPipelinesHandler handler = serverConfiguration.listPipelinesHandler(dataPrepper);
+        final ListPipelinesHandler handler = serverConfiguration.listPipelinesHandler(pipelinesProvider);
 
         assertThat(handler, isA(ListPipelinesHandler.class));
     }
 
     @Test
     public void testGivenValidInputWithAuthenticatorThenServerListContextCreated() {
-        final DataPrepper dataPrepper = mock(DataPrepper.class);
+        final PipelinesProvider pipelinesProvider = mock(PipelinesProvider.class);
 
-        final ListPipelinesHandler handler = serverConfiguration.listPipelinesHandler(dataPrepper);
+        final ListPipelinesHandler handler = serverConfiguration.listPipelinesHandler(pipelinesProvider);
 
         assertThat(handler, isA(ListPipelinesHandler.class));
     }

--- a/data-prepper-main/src/main/java/org/opensearch/dataprepper/ContextManager.java
+++ b/data-prepper-main/src/main/java/org/opensearch/dataprepper/ContextManager.java
@@ -12,19 +12,10 @@ import org.springframework.core.env.SimpleCommandLinePropertySource;
 
 /**
  * @since 1.3
- * <p>Creates Spring {@link org.springframework.context.ApplicationContext} hierarchy for Dependency Injection with limited visibility.</p>
- * <p>
- *     Application Context Hierarchy
- *         Public Application Context<br>
- *         ├─ Core Application Context<br>
- *         ├─ Shared Plugin Application Context<br>
- *             ├─ Plugin Isolated Application Context<br>
- * </p>
  */
-public class ContextManager {
+public class ContextManager extends AbstractContextManager {
     private static final Logger LOG = LoggerFactory.getLogger(ContextManager.class);
-
-    private final AnnotationConfigApplicationContext coreApplicationContext;
+    private final SimpleCommandLinePropertySource commandLinePropertySource;
 
     /**
      * @since 1.3
@@ -34,26 +25,11 @@ public class ContextManager {
      */
     public ContextManager(final String ... args) {
         LOG.trace("Reading args");
-        final SimpleCommandLinePropertySource commandLinePropertySource = new SimpleCommandLinePropertySource(args);
-
-        final AnnotationConfigApplicationContext publicApplicationContext = new AnnotationConfigApplicationContext();
-        publicApplicationContext.scan("org.opensearch.dataprepper.expression");
-        publicApplicationContext.refresh();
-
-        coreApplicationContext = new AnnotationConfigApplicationContext();
-        coreApplicationContext.setParent(publicApplicationContext);
-        coreApplicationContext.getEnvironment().getPropertySources().addFirst(commandLinePropertySource);
-        coreApplicationContext.register(DataPrepperExecute.class);
-
-        coreApplicationContext.refresh();
+        commandLinePropertySource = new SimpleCommandLinePropertySource(args);
     }
 
-    /**
-     * @since 1.3
-     * Retrieves the instance of {@link DataPrepper} singleton
-     * @return {@link DataPrepper} instance
-     */
-    public DataPrepper getDataPrepperBean() {
-        return coreApplicationContext.getBean(DataPrepper.class);
+    @Override
+    protected void preRefreshCoreApplicationContext(final AnnotationConfigApplicationContext coreApplicationContext) {
+        coreApplicationContext.getEnvironment().getPropertySources().addFirst(commandLinePropertySource);
     }
 }

